### PR TITLE
fix: method naming conflicts

### LINKED ISSUE #121: Standalone contract function name conflicting issue
If user write the contract method name same as one of C function, it would case build or execute fail, depending on which is conflict. Need to figure out a way to address this kind of confliction

### PR DESCRIPTION
**Pull request by osalkanovic**
Issue: #121 

Here is working example how to prevent same functions naming as c functions. 
I tried to move functions to helper, but unsuccessfully to import helper inside babel transformer, if someone can assist it would be good :))

### CODE REVIEW

File: src/build-tools/near-bindgen-exporter.js [hunk 1]
```diff
@@ -25,6 +28,7 @@ export default function () {
             }
           }
           for (let method of Object.keys(contractMethods)) {
+            validateMethod(method, cFunctions)
```
Comment (ailisp) [inline src/build-tools/near-bindgen-exporter.js:31]: This validating pattern is cool!

File: src/build-tools/near-bindgen-exporter.js [hunk 2]
```diff
@@ -67,3 +71,19 @@ export default function () {
     }
   };
 }
+
+
+function getCFunctions(){
+  const builderRaw = readFileSync(path.resolve('../cli/builder/builder.c'), 'utf-8')
+  const regex = new RegExp(/(?<=JS_SetPropertyStr\(ctx, env, ").+?(?=", JS_NewCFunction\(ctx,)/gm)
```
Comment (ailisp) [inline src/build-tools/near-bindgen-exporter.js:78]: There are more functions to filter out, basically every C function that's included in builder.c, such as `strncpy`, so this is quite challenging to figure out all C function names. I wonder whether one these directions can be easier:
- when confliction happens, can we collect the C compiler's error/warning and indicate the user to rename the contract method
- can we iterate contract method names like you did, then for any name confliction, use some C code-transformation / compiler tool to rename C functions to some random names. For example, a developer can write a contract method named `strncpy` (:joy:), and we let the tool to rename original `strncpy` to `___strncpy`

Comment (osalkanovic) [reply src/build-tools/near-bindgen-exporter.js:78]: Wooow, yes. Why not, I will update PR

**Comment by volovyks** [comment]
The reason why you can not import these functions from utils is that Babel not allowing imports without .js file extension.
https://stackoverflow.com/questions/67996772/babel-not-allowing-imports-without-js-file-extension

You are importing it from `../utils`, but Babel expects `../unils.js` from `/lib` folder. You can try to apply the suggested solution, but it is not a big deal.

**Comment by ailisp** [comment]
@osalkanovic Looks like a `yarn build` to update the built artifacts is required to pass CI

**Comment by osalkanovic** [comment]
@ailisp can you re-run CI?

**Comment by osalkanovic** [comment]
@ailisp first step is resolved
output from `groupCFunctions` is 
```
 'JS_NewCustomContext' => [ { x: 5, y: 18 }, { x: 33, y: 10 } ],
  'JS_NewContextRaw' => [ { x: 7, y: 19 } ],
  'JS_AddIntrinsicBaseObjects' => [ { x: 10, y: 2 } ],
  'JS_AddIntrinsicDate' => [ { x: 11, y: 2 } ],
  'JS_AddIntrinsicEval' => [ { x: 12, y: 2 } ],
  'JS_AddIntrinsicStringNormalize' => [ { x: 13, y: 2 } ],
  'JS_AddIntrinsicRegExp' => [ { x: 14, y: 2 } ],
  'JS_AddIntrinsicJSON' => [ { x: 15, y: 2 } ],
  'JS_AddIntrinsicProxy' => [ { x: 16, y: 2 } ],
  'JS_AddIntrinsicMapSet' => [ { x: 17, y: 2 } ],
  'JS_AddIntrinsicTypedArrays' => [ { x: 18, y: 2 } ],
  'JS_AddIntrinsicPromise' => [ { x: 19, y: 2 } ],
  'JS_AddIntrinsicBigInt' => [ { x: 20, y: 2 } ],
  'DEFINE_NEAR_METHOD' => [ { x: 24, y: 8 } ],
  '__attribute__' => [ { x: 25, y: 15 } ],
  'export_name' => [ { x: 25, y: 30 } ],
  'JS_NewRuntime' => [ { x: 32, y: 9 } ],
  'js_add_near_host_functions' => [ { x: 34, y: 4 }, { x: 960, y: 12 } ],
  'js_load_module_binary' => [ { x: 35, y: 14 } ],
  'JS_GetProperty' => [ { x: 36, y: 14 } ],
  'JS_NewAtom' => [ { x: 36, y: 43 } ],
  'JS_Call' => [ { x: 37, y: 13 }, { x: 1030, y: 8 } ],
  'JS_IsException' => [ { x: 38, y: 8 } ],
  'JS_GetException' => [ { x: 39, y: 14 } ],
  'JS_GetPropertyStr' => [ { x: 40, y: 22 }, { x: 41, y: 20 } ],
  'JS_ToCStringLen' => [
    { x: 42, y: 24 },  { x: 43, y: 22 },
    { x: 488, y: 15 }, { x: 512, y: 13 },
    { x: 546, y: 19 }, { x: 547, y: 20 },
    { x: 572, y: 19 }, { x: 573, y: 20 },
    { x: 606, y: 19 }, { x: 621, y: 19 },
....
....
```

Now we have coordinates where function start in builder.c so we can easily include __ in function name :D 

I have to also found coordinates for regex match functions :D 
I will update PR with step 2 in next days :) 

cc @volovyks can you assist with moving functions to utils?

**Comment by osalkanovic** [comment]
CI should fail because I have c function in smart contract

**Comment by ailisp** [comment]
@osalkanovic Seems that your collectCFunction/groupCFunction implemented a great algorithm that parsed all c function used in builder.c! However it doesn't collect the names in `#include` files. The C programming language is quite low level, there's no namespace or module concept, the `#Include` is equivalent to litterally copy-paste such file in place. So the function defines in `#include` (unless `static` ones, which are file-local names), will also conflict. 

I suggest to look for some existing unix tool that recursively process C includes, for example `ctags` might work. The good news about the include files are they are almost fixed and we almost never change them, so a one time figure out the reserved function names from include files is good enough. For builder.c it can be changed in future near-sdk-js and your automatic groupCFunction is nice!

**Comment by ailisp** [comment]
Fixed in #346 . @osalkanovic Appreciate your contributions to near-sdk-js! We are now accepting proposals and give fundings to improve NEAR developer tools, including JS SDK. If you have an idea about near-sdk-js, create an `Idea` post labeled with near-sdk-js and reply to existing ideas with `Solution` post with your proposal and financial ask. We'll discuss them on a regular basis and fund proposals that pass the vote in our work group. Then you can start work on it with funding.

The place to submit idea / solution: https://devgovgigs.near.social

**Pull request closed without merging**